fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/EnvironmentRenderers.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.rendering.v1;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.render.LightmapTextureManager;
+import net.minecraft.client.render.SkyProperties;
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.util.math.Matrix4f;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.mixin.client.rendering.SkyPropertiesAccessor;
+
+/**
+ * Environmental renderers render world specific visuals of a world.
+ * They may be used to render the sky, weather, or clouds.
+ * The {@link SkyProperties} is the vanilla environmental renderer.
+ */
+@Environment(EnvType.CLIENT)
+public final class EnvironmentRenderers {
+	private static final Map<RegistryKey<World>, SkyRenderer> SKY_RENDERERS = new IdentityHashMap<>();
+	private static final Map<RegistryKey<World>, CloudRenderer> CLOUD_RENDERERS = new HashMap<>();
+	private static final Map<RegistryKey<World>, WeatherRenderer> WEATHER_RENDERERS = new HashMap<>();
+
+	/**
+	 * Registers a custom sky renderer for a DimensionType.
+	 *
+	 * <p>This overrides Vanilla's sky rendering.
+	 * @param key A RegistryKey for your DimensionType
+	 * @param renderer A {@link SkyRenderer} implementation
+	 * @param override Should try to override current SkyRenderer if it exists
+	 * @return if the new renderer was accepted
+	 */
+	public static boolean registerSkyRenderer(RegistryKey<World> key, SkyRenderer renderer, boolean override) {
+		Objects.requireNonNull(key);
+		Objects.requireNonNull(renderer);
+
+		if (!override && SKY_RENDERERS.containsKey(key)) {
+			return false;
+		} else {
+			SKY_RENDERERS.put(key, renderer);
+			return true;
+		}
+	}
+
+	/**
+	 * Registers a custom weather renderer for a DimensionType.
+	 *
+	 * <p>This overrides Vanilla's weather rendering.
+	 * @param key A RegistryKey for your Dimension Type
+	 * @param renderer A {@link WeatherRenderer} implementation
+	 * @param override Should try to override current SkyRenderer if it exists
```

Comment:
Its not ideal that only one mod can register a renderer, I guess a mod could chain the renderers by calling `get*Renderer` first? Is this something that could/should be handled better? Mod load order is none deremisitic so its technically possible for your sky to change each time to load the game?

Could be something we just accept.

Reply:
If anyone has suggestion, feel free to propose.